lib/internal/loader/Loader.js

```diff
@@ -26,47 +28,77 @@ function getBase() {
 }
 
 class Loader {
-  constructor(base = getBase()) {
+  constructor({ resolve = ModuleRequest.resolve, dynamicInstantiate } = {},
+              base = getBase()) {
     this.moduleMap = new ModuleMap();
     if (typeof base !== 'undefined' && base instanceof URL !== true) {
       throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'base', 'URL');
     }
     this.base = base;
+    this.resolver = resolve.bind(null);
+    this.dynamicInstantiate = dynamicInstantiate;
   }
 
-  async resolve(specifier) {
-    const request = resolveRequestUrl(this.base, specifier);
-    if (request.url.protocol !== 'file:') {
-      throw new errors.Error('ERR_INVALID_PROTOCOL',
-                             request.url.protocol, 'file:');
+  async resolve(specifier, parentURLOrString = this.base) {
+    if (typeof parentURLOrString === 'string') {
+      parentURLOrString = new URL(parentURLOrString);
+    } else if (parentURLOrString instanceof URL === false) {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE',
+                                 'parentURLOrString', 'URL');
```

Comment:
This doesn't really address Domenic's concern of a mutable `URL` object being passed around. Instead you can do `if (typeof parentURLOrString !== 'string' && !(parentURLOrString instanceof URL)) throw ...; const url = new URL(parentURLOrString);`.

Reply:
Or just do ToString on any incoming argument like every ES and web API do.

Reply:
I think the Node code precedent here is to use template strings to coerce:
```js
`${url}`
```

Reply:
This resolve isn't exposed publicly, it's the `this.resolver` call just below that represents the hookable API. Will add a guard to that.